### PR TITLE
feat(graphql): add support for GraphQL queries alongside existing REST API

### PR DESCRIPTION
**Pull request by tamirGer**

### CODE REVIEW

File: src/testimonials/testimonials.controller.ts [hunk 1]
```diff
@@ -83,12 +83,12 @@ export class TestimonialsController {
...
-    description: SWAGGER_DESC_GET_TESTIMONIALS_ON_SQL_QUERY,
+    description: API_DESC_GET_TESTIMONIALS_ON_SQL_QUERY,
   })
   @ApiOkResponse({
     type: String,
   })
-  async getCount(@Query('query') query: string): Promise<string> {
+  async getCount(@Query('query') query: string): Promise<number> {
     this.logger.debug('Get count of testimonials.');
     return await this.testimonialsService.count(query);
```
Comment (derevnjuk) [inline src/testimonials/testimonials.controller.ts:93]: Avoid redundant async/await statements if possible

```suggestion
  getCount(@Query('query') query: string): Promise<number> {
    this.logger.debug('Get count of testimonials.');
    return this.testimonialsService.count(query);
```

File: src/testimonials/testimonials.resolver.ts [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+import { TestimonialDto } from './api/TestimonialDto';
+import { TestimonialsService } from './testimonials.service';
+import { CreateTestimonialRequest } from './api/CreateTestimonialRequestREST';
+import {
+  API_DESC_CREATE_TESTIMONIAL,
+  API_DESC_GET_TESTIMONIALS,
+  API_DESC_GET_TESTIMONIALS_ON_SQL_QUERY,
+} from './testimonials.controller.api.desc';
+
+@Resolver((of) => Testimonial)
```
Comment (derevnjuk) [inline src/testimonials/testimonials.resolver.ts:20]: Remove unused parameters  

```suggestion
@Resolver(() => Testimonial)
```

File: src/testimonials/testimonials.resolver.ts [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+export class TestimonialsResolver {
+  private readonly logger = new Logger(TestimonialsResolver.name);
+
+  constructor(private readonly testimonialsService: TestimonialsService) {}
+
+  @Query((returns) => [Testimonial], {
```
Comment (derevnjuk) [inline src/testimonials/testimonials.resolver.ts:26]:
```suggestion
  @Query(() => [Testimonial], {
```

File: src/testimonials/testimonials.resolver.ts [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    );
+  }
+
+  @Query((returns) => Int, {
+    description: API_DESC_GET_TESTIMONIALS_ON_SQL_QUERY,
+  })
+  async testimonialsCount(@Args('query') query: string): Promise<number> {
+    this.logger.debug('Get count of testimonials.');
+    return await this.testimonialsService.count(query);
+  }
```
Comment (derevnjuk) [inline src/testimonials/testimonials.resolver.ts:42]:
```suggestion
  testimonialsCount(@Args('query') query: string): Promise<number> {
    this.logger.debug('Get count of testimonials.');
    return this.testimonialsService.count(query);
  }
```

File: src/testimonials/testimonials.resolver.ts [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    description: API_DESC_GET_TESTIMONIALS,
+  })
+  async allTestimonials(): Promise<Testimonial[]> {
+    this.logger.debug('Get all testimonials.');
+    return (await this.testimonialsService.findAll()).map<TestimonialDto>(
+      TestimonialDto.covertToApi,
+    );
+  }
+
+  @Query((returns) => Int, {
```
Comment (derevnjuk) [inline src/testimonials/testimonials.resolver.ts:36]:
```suggestion
  @Query(() => Int, {
```

File: src/testimonials/testimonials.resolver.ts [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    description: API_DESC_GET_TESTIMONIALS_ON_SQL_QUERY,
+  })
+  async testimonialsCount(@Args('query') query: string): Promise<number> {
+    this.logger.debug('Get count of testimonials.');
+    return await this.testimonialsService.count(query);
+  }
+
+  @Mutation((returns) => Testimonial, {
```
Comment (derevnjuk) [inline src/testimonials/testimonials.resolver.ts:44]:
```suggestion
  @Mutation(() => Testimonial, {
```

File: src/testimonials/testimonials.resolver.ts [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+      return TestimonialDto.covertToApi(
+        await this.testimonialsService.createTestimonial(
+          testimonialRequest.message,
+          testimonialRequest.name,
+          testimonialRequest.title,
+        ),
+      );
+    } catch {
+      throw InternalServerErrorException;
+    }
```
Comment (derevnjuk) [inline src/testimonials/testimonials.resolver.ts:63]: You should throw only instances of error (avoid passing primitive or other values to the `throw` function). Anyway, in this particular case, you can remove try/catch block entirely
```suggestion
    this.logger.debug('Create testimonial.');
    return TestimonialDto.covertToApi(
      await this.testimonialsService.createTestimonial(
        testimonialRequest.message,
        testimonialRequest.name,
        testimonialRequest.title,
      ),
    );
```

File: src/products/products.resolver.ts [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+    const products = await this.productsService.findLatest(3);
+    return products.map((p: Product) => new ProductDto(p));
+  }
+
+  @Mutation((returns) => Boolean, {
+    description: API_DESC_GET_VIEW_PRODUCT,
+  })
+  async viewProduct(
+    @Args('productName') productName: string,
+  ): Promise<Boolean> {
```
Comment (derevnjuk) [inline src/products/products.resolver.ts:40]:
```suggestion
  ): Promise<boolean> {
```

File: src/products/products.resolver.ts [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+
+  @Query((returns) => [Product], {
+    description: API_DESC_GET_LATEST_PRODUCTS,
+  })
+  async latestProducts(): Promise<Product[]> {
+    const products = await this.productsService.findLatest(3);
+    return products.map((p: Product) => new ProductDto(p));
+  }
+
+  @Mutation((returns) => Boolean, {
```
Comment (derevnjuk) [inline src/products/products.resolver.ts:35]:
```suggestion
  @Mutation(() => Boolean, {
```

File: src/products/products.resolver.ts [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+
+  @Query((returns) => [Product], { description: API_DESC_GET_PRODUCTS })
+  @UseGuards(AuthGuard)
+  @JwtType(JwtProcessorType.RSA)
+  async allProducts(): Promise<Product[]> {
+    const allProducts = await this.productsService.findAll();
+    return allProducts.map((p: Product) => new ProductDto(p));
+  }
+
+  @Query((returns) => [Product], {
```
Comment (derevnjuk) [inline src/products/products.resolver.ts:27]:
```suggestion
  @Query(() => [Product], {
```

File: src/products/products.resolver.ts [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+  API_DESC_GET_LATEST_PRODUCTS,
+  API_DESC_GET_PRODUCTS,
+  API_DESC_GET_VIEW_PRODUCT,
+} from './products.controller.api.desc';
+
+@Resolver((of) => Product)
+export class ProductsResolver {
+  constructor(private readonly productsService: ProductsService) {}
+
+  @Query((returns) => [Product], { description: API_DESC_GET_PRODUCTS })
```
Comment (derevnjuk) [inline src/products/products.resolver.ts:19]:
```suggestion
  @Query(() => [Product], { description: API_DESC_GET_PRODUCTS })
```

File: src/products/products.resolver.ts [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+import { Product } from './api/product.model';
+import { ProductDto } from './api/ProductDto';
+import { ProductsService } from './products.service';
+import {
+  API_DESC_GET_LATEST_PRODUCTS,
+  API_DESC_GET_PRODUCTS,
+  API_DESC_GET_VIEW_PRODUCT,
+} from './products.controller.api.desc';
+
+@Resolver((of) => Product)
```
Comment (derevnjuk) [inline src/products/products.resolver.ts:15]:
```suggestion
@Resolver(() => Product)
```

File: src/testimonials/testimonials.controller.ts (hunk 1 above)
Comment (tamirGer) [reply src/testimonials/testimonials.controller.ts:93]: As discussed, we'll do this for all controllers in a separate PR.

**Review by derevnjuk — APPROVED** [review]

**Pull request merged**